lua/firvish/config.lua

```diff
@@ -0,0 +1,189 @@
+                        vim.cmd "edit firvish://menu"
+                    end,
+                },
+            },
+        },
+        job = preview_mappings,
+        ["job-list"] = {
+            n = {
+                ["<enter>"] = {
+                    function()
+                        local line = vim.fn.line "."
+                        local lines = vim.api.nvim_buf_get_var(0, "firvish_job_list_additional_lines")
+                        require("firvish.job_control").preview_job_output(lines[line].job_id)
+                    end,
+                },
+                dd = {
+                    function()
+                        require("firvish.job_control").delete_job_from_history(false)
+                    end,
+                },
+                E = {
+                    function()
+                        local line = vim.fn.line "."
+                        local lines = vim.b.firvish_job_list_additional_lines
+                        require("firvish.job_control").echo_job_output(
+                            lines[line].job_id,
+                            math.max(vim.v.count, 1) * -1
+                        )
+                    end,
+                },
+                P = {
+                    function()
+                        local line = vim.fn.line "."
+                        local lines = vim.api.nvim_buf_get_var(0, "firvish_job_list_additional_lines")
+                        require("firvish.job_control").preview_job_output(lines[line].job_id)
+                    end,
+                },
+                R = {
+                    function()
+                        require("firvish.job_control").refresh_job_list_window()
+                    end,
+                },
+                S = {
+                    function()
+                        require("firvish.job_control").stop_job()
+                    end,
+                },
+            },
+        },
+        ["job-output"] = {
+            n = {
+                gb = {
+                    function()
+                        require("firvish.job_control").go_back_from_job_output()
+                    end,
+                },
+            },
+        },
+        menu = {
+            n = {
+                ["<enter>"] = {
+                    function()
+                        require("firvish.menu").open_item(vim.fn.line ".")
+                    end,
+                },
+                R = {
+                    function()
+                        require("firvish.menu").refresh_menu()
+                    end,
+                },
+            },
+        },
+    },
+}
+
+M.apply_mappings = function(map)
+    local config = M.config
+    for mode, maps in pairs(config.keymaps[map]) do
+        for lhs, opts in pairs(maps) do
+            if opts then
+                table.insert(opts, 1, lhs)
+                keymap.make_mapper(mode, { buffer = 0, noremap = true, silent = true }, opts)
+            end
+        end
+    end
```

Comment:
What do you think about this way of doing it? I prefer it for two reasons: (1) it's less code and (2) it lets you customize the opts pass to the underlying ?map function. So if someone wanted to they could;
```lua
require("firvish").setup {
  keymaps = {
    buffers = {
      n = {
        f = { function() ... end, expr = true }
      }
    }
  }
}
```

Reply:
I think it looks good. The only thing that got my attention is it's not very clear whether a mapping is done for a buffer or globally. Making that more explicit would be better.

Let me know when you think it's ready to merge and I'll do so.

Reply:
Hmm. Turns out it actually doesn't work, which I don't quite understand. It works the *first* time but not on subsequent enters into the various firvish buffers (firvish-job-list for example). Which is weird because nnoremap just calls make_mapper. But it works fine when we call nnoremap directly, but not make_mapper directly. Very odd. Anyways, I reverted it for now and will see if I can debug it further some other time.